internal/controller/agenttoken_controller.go

```diff
@@ -0,0 +1,421 @@
+		r.Recorder.Eventf(&t.instance, corev1.EventTypeWarning, "RemoveAgentToken", "Failed to remove finalizer %s", agentTokenFinalizer)
+	}
+
+	return err
+}
+
+func (r *AgentTokenReconciler) getAgentPoolIDByName(ctx context.Context, t *agentTokenInstance) (*tfc.AgentPool, error) {
+	spec := t.instance.Spec.AgentPool.Name
+
+	listOpts := &tfc.AgentPoolListOptions{
+		Query: spec,
+		ListOptions: tfc.ListOptions{
+			PageSize: maxPageSize,
+		},
+	}
+	for {
+		agentPoolIDs, err := t.tfClient.Client.AgentPools.List(ctx, t.instance.Spec.Organization, listOpts)
+		if err != nil {
+			return nil, err
+		}
+		for _, a := range agentPoolIDs.Items {
+			if a.Name == spec {
+				return a, nil
+			}
+		}
+		if agentPoolIDs.NextPage == 0 {
+			break
+		}
+		listOpts.PageNumber = agentPoolIDs.NextPage
+	}
+
+	return nil, fmt.Errorf("agent pool ID not found for agent pool name %q", spec)
+}
+
+func (r *AgentTokenReconciler) getAgentPoolID(ctx context.Context, t *agentTokenInstance) (*tfc.AgentPool, error) {
+	specAgentPool := t.instance.Spec.AgentPool
+
+	if specAgentPool.Name != "" {
+		t.log.Info("Reconcile Agent Pool", "msg", "getting agent pool ID by name")
+		return r.getAgentPoolIDByName(ctx, t)
+	}
+
+	t.log.Info("Reconcile Agent Pool", "msg", "getting agent pool ID from the spec.AgentPool.ID")
+
+	return t.tfClient.Client.AgentPools.Read(ctx, specAgentPool.ID)
+}
+
+func (r *AgentTokenReconciler) getAgentPool(ctx context.Context, t *agentTokenInstance) (*tfc.AgentPool, error) {
+	spec := t.instance.Spec.AgentPool
+
+	if spec.Name != "" {
+		t.log.Info("Reconcile Agent Token", "msg", "getting agent pool by name")
+		return r.getAgentPoolIDByName(ctx, t)
+	}
+
+	t.log.Info("Reconcile Agent Token", "msg", "getting agent pool by ID")
+	return r.getAgentPoolID(ctx, t)
+}
+
+func (r *AgentTokenReconciler) updateStatusAgentPool(ctx context.Context, t *agentTokenInstance) error {
+	pool, err := r.getAgentPool(ctx, t)
+	if err != nil {
+		return err
+	}
+
+	t.instance.Status.AgentPool = &appv1alpha2.AgentPoolRef{
+		ID:   pool.ID,
+		Name: pool.Name,
+	}
+
+	return nil
+}
+
+func (r *AgentTokenReconciler) listAgentTokens(ctx context.Context, t *agentTokenInstance) (map[string]string, error) {
+	if t.instance.Status.AgentPool == nil {
+		if err := r.updateStatusAgentPool(ctx, t); err != nil {
+			return nil, err
+		}
+	}
+	m := make(map[string]string)
+	tokens, err := t.tfClient.Client.AgentTokens.List(ctx, t.instance.Status.AgentPool.ID)
+	if err == tfc.ErrResourceNotFound {
+		if err := r.updateStatusAgentPool(ctx, t); err != nil {
+			return nil, err
+		}
+		tokens, err := t.tfClient.Client.AgentTokens.List(ctx, t.instance.Status.AgentPool.ID)
+		if err != nil {
+			return nil, err
+		}
+		for _, token := range tokens.Items {
+			m[token.ID] = token.Description
+		}
+		return m, err
+	}
+	for _, token := range tokens.Items {
+		m[token.ID] = token.Description
+	}
```

Comment:
Could perhaps make this logic a bit more concise, like this:

```golang
tokens, err := t.tfClient.Client.AgentTokens.List(ctx, t.instance.Status.AgentPool.ID)
if err == tfc.ErrResourceNotFound {
	if err := r.updateStatusAgentPool(ctx, t); err != nil {
		return nil, err
	}
	tokens, err = t.tfClient.Client.AgentTokens.List(ctx, t.instance.Status.AgentPool.ID)
}
if err != nil {
	return nil, err
}

for _, token := range tokens.Items {
	m[token.ID] = token.Description
}
```

Reply:
There are several places with similar logic; in the future, we should switch to a more generic retry approach.